keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane/README.md

```diff
@@ -0,0 +1,1077 @@
+- /etc/kubernetes/pki/sa.key
+
+We propose that files shared by `kube-controller-manager` and `kube-apiserver` be readable by a particular `GID` and we set the `supplementalGroups` in the [PodSecurityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podsecuritycontext-v1beta1-policy) of the Pod to this GID for `kube-apiserver` and `kube-controller-manager`.
+
+For instance lets consider that `kubeadm` grant group 2100 read permissions to /etc/kubernetes/pki/ca.crt
+
+Then we update the `kube-apiserver` manifest as follows:-
+
+```yaml
+
+apiVersion: v1
+kind: Pod
+metadata:
+  annotations:
+    kubeadm.kubernetes.io/kube-apiserver.advertise-address.endpoint: 172.17.0.2:6443
+  creationTimestamp: null
+  labels:
+    component: kube-apiserver
+    tier: control-plane
+  name: kube-apiserver
+  namespace: kube-system
+spec:
+  securityContext:
+    supplementalGroups:
+    - 2100  # this value is only for an example and is not the id we plan to use.
+  containers:
+  - name: kube-apiserver
+    command:
+    - kube-apiserver
+    - --advertise-address=172.17.0.2
+    - --client-ca-file=/etc/kubernetes/pki/ca.crt
+    ... # omitted to save space 
+    securityContext:
+      runAsUser: 2000  # this value is only an example and is not the id we plan to use.
+      runAsGroup: 2000  # this value is only an example and is not the id we plan to use.
+      allowPrivilegeEscalation: false
+      seccompProfile:
+        type: runtime/default
+      capabilities:
+        drop:
+        - ALL
+    image: k8s.gcr.io/kube-apiserver:v1.21.0-beta.0.368_9850bf06b571d5-dirty
+    ... # omitted to save space
+    volumeMounts:
+    ... # omitted to save space
+    - mountPath: /etc/kubernetes/pki
+      name: k8s-certs
+      readOnly: true
+    ... # omitted to save space
+```
+
+Similarly `kube-controller-manager`'s manifest would like:-
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  creationTimestamp: null
+  labels:
+    component: kube-controller-manager
+    tier: control-plane
+  name: kube-controller-manager
+  namespace: kube-system
+spec:
+  securityContext:
+    supplementalGroups:
+    - 2100
+  containers:
+  - name: kube-controller-manager
+    command:
+    - kube-controller-manager
+    - --cluster-signing-cert-file=/etc/kubernetes/pki/ca.crt
+    ...
+    # omitted to save space.
+    securityContext:
+      runAsUser: 2001  # this value is only an example and is not the id we plan to use.
+      runAsGroup: 2001 # this value is only an example and is not the id we plan to use.
+      allowPrivilegeEscalation: false
+      seccompProfile:
+        type: runtime/default
+      capabilities:
+        drop:
+        - ALL
+    image: k8s.gcr.io/kube-controller-manager:v1.21.0-beta.0.368_9850bf06b571d5-dirty
+    ... # omitted to save space.
+    volumeMounts:
+    - mountPath: /etc/kubernetes/pki
+      name: k8s-certs
+      readOnly: true
+    ... # omitted to save space. 
+```
+<br/>
```

Comment:
nit: i noticed these BR tags in a few places. can we omit them?

Reply:
Done.